src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py

```diff
@@ -489,24 +489,38 @@ def __copy__(self) -> "SnowflakePlan":
                 referenced_ctes=self.referenced_ctes,
             )
 
-    def __deepcopy__(self, memodict={}) -> "SnowflakePlan":  # noqa: B006
-        copied_source_plan = (
-            copy.deepcopy(self.source_plan) if self.source_plan else None
-        )
+    def __deepcopy__(self, memodict=None) -> "SnowflakePlan":  # noqa: B006
+        if memodict is None:
+            memodict = {}
+
+        if (self_id := id(self)) in memodict:
```

Comment:
can we add a comment here that if the node has already been copied, return the copied node